lib/scholar/metrics.ex

```diff
@@ -481,6 +482,176 @@ defmodule Scholar.Metrics do
     (diff * diff) |> Nx.mean()
   end
 
+  @doc ~S"""
+  Computes area under the curve (AUC) using the trapezoidal rule.
+
+  This is a general function, given points on a curve.
+
+  ## Examples
+
+      iex> y = Nx.tensor([0, 0, 1, 1])
+      iex> pred = Nx.tensor([0.1, 0.4, 0.35, 0.8])
+      iex> distinct_value_indices = Scholar.Metrics.calculate_distinct_value_indices(pred)
+      iex> {fpr, tpr, _thresholds} = Scholar.Metrics.roc_curve(y, pred, distinct_value_indices)
+      iex> Scholar.Metrics.auc(fpr, tpr)
+      #Nx.Tensor<
+        f32
+        0.75
+      >
+  """
+  defn auc(x, y) do
+    check_shape(x, y)
+    dx = Nx.diff(x)
+
+    # 0 means x is neither increasing nor decreasing -> error
+    direction =
+      cond do
+        Nx.all(dx <= 0) -> -1
+        Nx.all(dx >= 0) -> 1
+        true -> Nx.tensor(:nan, type: to_float_type(y))
+      end
+
+    direction * Trapezoidal.trapezoidal(y, x)
+  end
+
+  @doc ~S"""
+  It's a helper function for `Scholar.Metrics.roc_curve` and `Scholar.Metrics.roc_auc_score` functions.
+  You should call it and use as follows:
+
+      distinct_value_indices = Scholar.Metrics.calculate_distinct_value_indices(scores)
+      {fpr, tpr, thresholds} = Scholar.Metrics.roc_curve(y_true, scores, distinct_value_indices, weights)
+  """
+  def calculate_distinct_value_indices(y_score) do
```

Comment:
Let's rename this function to `distinct_value_indices` (no need to `calculate`). I have also noticed that none of our tests have duplicate values, so this behaviour itself is not being tested.